protocols/upnp/src/provider.rs

```diff
@@ -0,0 +1,173 @@
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{error::Error, fmt, net::Ipv4Addr};
+
+use crate::{
+    behaviour::{GatewayEvent, GatewayRequest},
+    Config,
+};
+use async_trait::async_trait;
+use futures::channel::mpsc::{Receiver, Sender};
+
+/// Protocols available for port mapping.
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+pub(crate) enum Protocol {
+    Tcp,
+    Udp,
+}
+
+impl fmt::Display for Protocol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Protocol::Tcp => f.write_str("tcp"),
+            Protocol::Udp => f.write_str("udp"),
+        }
+    }
+}
+
+/// Interface that interacts with the inner gateway by messages,
+/// `GatewayRequest`s and `GatewayEvent`s.
+pub struct Gateway {
+    pub(crate) sender: Sender<GatewayRequest>,
+    pub(crate) receiver: Receiver<GatewayEvent>,
+    pub(crate) addr: Ipv4Addr,
+}
+
+/// Abstraction to allow for compatibility with various async runtimes.
+#[async_trait]
+pub trait Provider {
+    async fn search_gateway(config: Config) -> Result<Gateway, Box<dyn Error>>;
+}
+
+macro_rules! impl_provider {
+    ($impl:ident, $executor: ident, $gateway:ident, $protocol: ident, $options: ident) => {
+        use crate::behaviour::{GatewayRequest, GatewayEvent};
+        use super::{Gateway, Config, Protocol};
+
+        use async_trait::async_trait;
+        use std::error::Error;
+        use futures::{channel::mpsc, select, stream::FuturesUnordered, FutureExt, StreamExt, SinkExt};
+
+        impl From<Protocol> for $protocol {
+            fn from(protocol: Protocol) -> Self {
+                match protocol {
+                    Protocol::Tcp => $protocol::TCP,
+                    Protocol::Udp => $protocol::UDP,
+                }
+            }
+        }
+
+        #[async_trait]
+        impl super::Provider for $impl {
+            async fn search_gateway(config: Config) -> Result<super::Gateway, Box<dyn Error>> {
+                let options = $options {
+                    bind_addr: config.bind_addr,
+                    broadcast_address: config.broadcast_addr,
+                    timeout: config.timeout,
+                };
+                let gateway = $gateway::search_gateway(options).await?;
+                let addr = gateway.get_external_ip().await?;
+
+                let (events_sender, mut task_receiver) = mpsc::channel(5);
+                let (mut task_sender, events_queue) = mpsc::channel(0);
+
+                $executor::spawn(async move {
+                    let mut futs = FuturesUnordered::new();
```

Comment:
Following up on a different conversation in this pull request, namely on `FuturesUnordered` being unbounded, you could select some arbitrary limit. When that limit is reached you would stop polling `task_receiver.select_next_some()` until `FuturesUnordered::len()` is again below the limit.

Reply:
> the problem with that is that if doing it sequentially, while waiting for the limit to be reached, we won't stop polling `task_receiver.select_next_some()` and therefore we don't communicate open ports. But we can actually do it sequentially, not use `FuturesUnordered` and for each request for port mapping we wait for it to complete before returning it's result, and the only buffer is the channel for the `task_receiver`. Subsequentially the scenario you described on [#4156 (comment)](https://github.com/libp2p/rust-libp2p/pull/4156#discussion_r1264877548) doesn't occurr. Ptal :)

Good idea. Implementation looks good to me.